mllib/src/main/scala/org/apache/spark/mllib/stat/test/ADTest.scala

```diff
@@ -0,0 +1,269 @@
+   * @param makeDist `() => ADTheoreticalDist` a function to create a class that extends the
+   *                ADTheoreticalDist trait, which requires various methods, used in creating 1
+   *                object per partition
+   * @param n `Double` the total size of the data sample
+   * @return `Iterator[(Double, Double, Double)]` The first element corresponds to the
+   *        position-independent contribution to the AD statistic, the second is the value that must
+   *        be scaled by the number of elements in prior partitions and the third is the number of
+   *        elements in this partition
+   */
+  def calcPartAD(part: Iterator[Double], makeDist: () => ADTheoreticalDist, n: Double)
+    : Iterator[(Double, Double, Double)] = {
+      val dist = makeDist()
+      val initAcc = (0.0, 0.0, 0.0)
+      val pResult = part.zipWithIndex.foldLeft(initAcc) { case ((prevS, prevC, prevCt), (v, i)) =>
+        val y = dist.cdf(v)
+        val a = ln(y)
+        val b = ln(1 - y)
+        val unAdjusted = a * (2 * i + 1) + b * (2 * n - 2 * i - 1)
+        val adjConstant = a - b
+        (prevS + unAdjusted, prevC + adjConstant, prevCt + 1)
+      }
+    Array(pResult).iterator
+  }
+
+  /**
+   * Create a function to produce a distribution per partition.
+   * If the user provides parameters, the distribution is initialized with those values
+   * (note that the critical values still assume that the parameters were estimated from the data,
+   * the ability to provide the parameters is simply a convenience to avoid recomputing them
+   * when the user already has them available). If no parameters are provided, and the distribution
+   * is one of [normal, exponential], then the distribution is initialized with MLE values
+   * @param distName `String` name of distribution
+   * @param data `RDD[Double]` sample of data to analyze (and hence to use for parameter estimation
+   *            where possible)
+   * @param n `Double` size of data sample
+   * @param params `Double*` Initialization parameters for distribution
+   * @return `() => RealDistribution` function to create distribution object
+   */
+  def initDist(distName: String, data: RDD[Double], n: Double, params: Array[Double])
+    : () => ADTheoreticalDist = {
+    distName match {
+      case "norm" => {
+        val muHat = if (params.nonEmpty) params(0) else data.mean()
+        val sdHat = if (params.length > 1) {
+            params(1)
+          } else {
+            math.sqrt(data.map(x => math.pow(x - muHat, 2)).sum() / (n - 1))
+        }
+        () => new ADNormal(Array(muHat, sdHat))
+      }
+      case "exp" => {
+        val meanHat = if (params.nonEmpty) params(0) else data.mean()
+        () => new ADExponential(Array(meanHat))
+      }
+      case "gumbel" => {
+        if (params.length < 2) {
+          throw new Exception("Gumbel does not yet support parameter estimation. " +
```

Comment:
changed all the parameter checking for distributions to use require. Should the final case _ => be replaced as well?